src/ent/bit_rack.h

```diff
@@ -156,20 +206,208 @@ void bit_rack_div_mod(const BitRack *bit_rack, uint32_t divisor,
+  return *bit_rack * multiplier;
+#else
+  const uint64_t lowest_32 = bit_rack->low & 0xFFFFFFFF;
+  const uint64_t second_32 = bit_rack->low >> 32;
+  const uint64_t third_32 = bit_rack->high; // fits in 32 bits, no need to mask
+
+  const uint64_t low_product = lowest_32 * multiplier;
+  uint64_t carry = low_product >> 32;
+  const uint64_t lowest_result = low_product & 0xFFFFFFFF;
+
+  const uint64_t second_product = second_32 * multiplier + carry;
+  carry = second_product >> 32;
+  const uint64_t second_result = second_product & 0xFFFFFFFF;
+
+  const uint64_t high_result = third_32 * multiplier + carry;
+  const uint64_t low_result = (second_result << 32) | lowest_result;
+  return (BitRack){low_result, high_result};
+#endif
+}
+
+static inline void bit_rack_set_letter_count(BitRack *bit_rack, uint8_t ml,
+                                             uint8_t count) {
+  const int shift = ml * BIT_RACK_BITS_PER_LETTER;
+#if USE_INT128_INTRINSIC
+  *bit_rack &=
+      ~((unsigned __int128)((1 << BIT_RACK_BITS_PER_LETTER) - 1) << shift);
+  *bit_rack |= (unsigned __int128)count << shift;
+#else
+  uint64_t mask = ((uint64_t)1 << BIT_RACK_BITS_PER_LETTER) - 1;
+  if (shift < 64) {
+    bit_rack->low &= ~(mask << shift);
+    bit_rack->low |= ((uint64_t)count) << shift;
+  } else {
+    const int adjusted_shift = shift - 64;
+    bit_rack->high &= ~(mask << adjusted_shift);
+    bit_rack->high |= ((uint64_t)count) << adjusted_shift;
+  }
+#endif
+}
+
+static inline BitRack largest_bit_rack_for_ld(const LetterDistribution *ld) {
+  BitRack bit_rack = bit_rack_create_empty();
+  int letters_to_add = BOARD_DIM;
+  for (int ml = ld->size - 1; ml >= 1; ml--) {
+    int letter_count = ld->distribution[ml];
+    if (ml == ld->size - 1) {
+      letter_count += ld->distribution[BLANK_MACHINE_LETTER];
+    }
+    if (letters_to_add >= letter_count) {
+      bit_rack_set_letter_count(&bit_rack, ml, letter_count);
+      letters_to_add -= letter_count;
+    } else {
+      bit_rack_set_letter_count(&bit_rack, ml, letters_to_add);
+      return bit_rack;
+    }
+  }
+  return bit_rack;
+}
+
+static inline void bit_rack_add_letter(BitRack *bit_rack, uint8_t ml) {
+  const int shift = ml * BIT_RACK_BITS_PER_LETTER;
+#if USE_INT128_INTRINSIC
+  *bit_rack += (unsigned __int128)1 << shift;
+#else
+  if (shift < 64) {
+    bit_rack->low += (uint64_t)1 << shift;
+  } else {
+    bit_rack->high += (uint64_t)1 << (shift - 64);
+  }
+#endif
+}
+
+static inline void bit_rack_take_letter(BitRack *bit_rack, uint8_t ml) {
+  const int shift = ml * BIT_RACK_BITS_PER_LETTER;
+#if USE_INT128_INTRINSIC
+  *bit_rack -= (unsigned __int128)1 << shift;
+#else
+  if (shift < 64) {
+    bit_rack->low -= (uint64_t)1 << shift;
+  } else {
+    bit_rack->high -= (uint64_t)1 << (shift - 64);
+  }
+#endif
+}
+
+static inline void bit_rack_write_12_bytes(const BitRack *bit_rack,
+                                           uint8_t bytes[12]) {
+  // assumes little-endian
```

Comment:
Nit (only change if convenient): if possible try to remain endian-ambiguous and use compat/endian_conv.h if that helps.

Reply:
i believe the last common big endian architecture for personal computing was powerpc (i had one)
but it's good to be hygienic about this stuff, and actually having thought more about it, it wouldn't be hard to be compatible (but it would take some effort to emulate a big endian machine)
i can fix this up

Reply:
I have done the endian swap using existing endian_conv.h functions for 64 and 32 but I didn't add these 96 bit swaps there, they just seem overly specific to this use case, especially since it depends on whether BitRack is  __int128 or a struct